lib/puppet/type/file.rb

```diff
@@ -668,6 +668,7 @@ def perform_recursion(path)
       :links => self[:links],
       :recurse => (self[:recurse] == :remote ? true : self[:recurse]),
       :recurselimit => self[:recurselimit],
+      :source_permissions => self[:source_permissions],
```

Comment:
This looks correct to me. The agent needs to specify whether it wants to use them (or not), and the server should process the request based on the query parameters specified in the `file_metadatas` search request.

But this makes me wonder how it ever worked? I'm guessing the ruby implementation of the puppet file serving code mistakenly had a hardcoded assumption that `file_metadatas` search requests should always use source permissions?